src/compiler/checker.ts

```diff
@@ -4110,7 +4132,22 @@ namespace ts {
                 }
 
                 if (minArgumentCount < 0) {
-                    minArgumentCount = declaration.parameters.length;
+                    minArgumentCount = declaration.parameters.length - (thisType ? 1 : 0);
+                }
+                if (!thisType && compilerOptions.strictThis) {
+                    if (declaration.kind === SyntaxKind.FunctionDeclaration
+                        || declaration.kind === SyntaxKind.CallSignature
+                        || declaration.kind == SyntaxKind.FunctionExpression
+                        || declaration.kind === SyntaxKind.FunctionType) {
+                        thisType = voidType;
+                    }
+                    else if ((declaration.kind === SyntaxKind.MethodDeclaration || declaration.kind === SyntaxKind.MethodSignature)
+                        && (isClassLike(declaration.parent) || declaration.parent.kind === SyntaxKind.InterfaceDeclaration)) {
+                        thisType = declaration.flags & NodeFlags.Static ?
+                            getWidenedType(checkExpression((<ClassLikeDeclaration>declaration.parent).name)) :

```

Comment:
I tried `getTypeOfSymbol(getSymbolOfNode(...` first. According to my notes, I had trouble getting `typeof C` vs `C` in the static case.

But it looks like I was wrong. It works now. I'm pretty sure some other change I made was wrong at the time.